rust/main/utils/run-locally/src/sealevel/termination_invariant.rs

```diff
@@ -0,0 +1,82 @@
+use std::path::Path;
+
+use maplit::hashmap;
+
+use crate::{
+    config::Config,
+    fetch_metric,
+    invariants::{
+        relayer_balance_check, relayer_termination_invariants_met,
+        scraper_termination_invariants_met,
+    },
+    logging::log,
+    sealevel::{solana::*, SOL_MESSAGES_EXPECTED, SOL_MESSAGES_WITH_NON_MATCHING_IGP},
+    RELAYER_METRICS_PORT,
+};
+
+/// Use the metrics to check if the relayer queues are empty and the expected
+/// number of messages have been sent.
+#[allow(clippy::unnecessary_get_then_check)] // TODO: `rustc` 1.80.1 clippy issue
+pub fn termination_invariants_met(
+    config: &Config,
+    starting_relayer_balance: f64,
+    solana_cli_tools_path: &Path,
+    solana_config_path: &Path,
+) -> eyre::Result<bool> {
+    let sol_messages_expected = SOL_MESSAGES_EXPECTED;
+    let sol_messages_with_non_matching_igp = SOL_MESSAGES_WITH_NON_MATCHING_IGP;
+
+    // this is total messages expected to be delivered
+    let total_messages_expected = sol_messages_expected;
+    let total_messages_dispatched = total_messages_expected + sol_messages_with_non_matching_igp;
+
+    // Also ensure the counter is as expected (total number of messages), summed
+    // across all mailboxes.
+    let msg_processed_count = fetch_metric(
+        RELAYER_METRICS_PORT,
+        "hyperlane_messages_processed_count",
+        &hashmap! {},
+    )?
+    .iter()
+    .sum::<u32>();
+
+    let gas_payment_events_count = fetch_metric(
+        RELAYER_METRICS_PORT,
+        "hyperlane_contract_sync_stored_events",
+        &hashmap! {"data_type" => "gas_payments"},
+    )?
+    .iter()
+    .sum::<u32>();
```

Comment:
these can be refactored into a function that reads agent metrics across E2Es as well